### PR TITLE
unix: Add sendfile to platforms that support it

### PR DESCRIPTION
**Pull request by kamalmarhubi**

### CODE REVIEW

File: src/unix/bsd/freebsdlike/freebsd.rs [hunk 1]
```diff
@@ -4,7 +4,16 @@ pub type fsfilcnt_t = ::uint64_t;
...
+pub const SF_MNOWAIT: ::c_int = 0x00000002;
+pub const SF_SYNC: ::c_int = 0x00000004;
+pub const SF_NOCACHE: ::c_int =   0x00000010;
 
 extern {
     pub fn __error() -> *mut ::c_int;
 }
+
+#[inline]
+pub fn SF_FLAGS(rh: ::c_int, flags: ::c_int) -> ::c_int {
```
Comment (alexcrichton) [inline src/unix/bsd/freebsdlike/freebsd.rs]: Having never used sendfile before, much less on freebsd, is this macro necessary for the usefulness of the function? Macros are unfortunately harder to validate, and I'd probably prefer to leave it if out if possible. If it's needed for inclusion, though, should be fine!

File: src/unix/bsd/apple/mod.rs [hunk 2]
```diff
@@ -250,6 +250,12 @@ s! {
         pub l_type: ::c_short,
         pub l_whence: ::c_short,
     }
+    pub struct sf_hdtr {
```
Comment (alexcrichton) [inline src/unix/bsd/apple/mod.rs]: Typically each struct has a blank line between it and the previous one (here and below)

File: src/unix/notbsd/mod.rs [hunk 3]
```diff
@@ -616,6 +616,7 @@ extern {
     pub fn fstatfs(fd: ::c_int, buf: *mut statfs) -> ::c_int;
     pub fn memrchr(cx: *const ::c_void, c: ::c_int, n: ::size_t) -> *mut ::c_void;
     pub fn syscall(num: ::c_long, ...) -> ::c_long;
+    pub fn sendfile(out_fd: ::c_int, in_fd: ::c_int, offset: *mut off_t, count: ::size_t) -> ::ssize_t;
```
Comment (alexcrichton) [inline src/unix/notbsd/mod.rs]: Could you be sure to wrap this to 80 chars? (surrounding style)

**Comment by alexcrichton** [comment]
Thanks @kamalmarhubi! Looks good to me modulo some nitpicks, otherwise just waiting on some green CI

File: src/unix/bsd/freebsdlike/freebsd.rs (hunk 1 above)
Comment (kamalmarhubi) [inline src/unix/bsd/freebsdlike/freebsd.rs]: It's actually brand new in FreeBSD: https://svnweb.freebsd.org/base?view=revision&revision=293439

I imagine it's not absolutely needed, as the man pages document the use of the upper and lower 16 bits for `flags`. I've removed it for now.

File: src/unix/bsd/apple/mod.rs (hunk 2 above)
Comment (kamalmarhubi) [inline src/unix/bsd/apple/mod.rs]: Done.

**Comment by kamalmarhubi** [comment]
Fixed issues. Accidentally pushed an intermediate change, so canceling this build would speed things up: https://travis-ci.org/rust-lang-nursery/libc/builds/107141122

**Comment by kamalmarhubi** [comment]
@alexcrichton regarding verifying macros, at least for 32 bit values we could exhaustively check they are the same. It'd add at most a few seconds to the CI builds, which have a variance way above that anyway. I'd have to poke around the CI more to see how everything works, but do you think it makes sense?

**Comment by kamalmarhubi** [comment]
Urg more `*const` use where it isn't actually in the signature. Another couple of hours...

**Comment by kamalmarhubi** [comment]
Had missed an arg from FreeBSD. Confused by the Apple error, will look more tomorrow.

**Comment by alexcrichton** [comment]
Ah yeah we can always add it later if we need to, I figure for now we could just take the conservative route. 

In terms of verification I'd just want to read the headers myself, but other than that it seems fine.

**Comment by kamalmarhubi** [comment]
> Ah yeah we can always add it later if we need to, I figure for now we could just take the conservative route.
:+1:

**Comment by kamalmarhubi** [comment]
I think freebsd is fixed with latest, but I have yet to take a look at Apple.

**Comment by kamalmarhubi** [comment]
```
all.c:8115:54: error: use of undeclared identifier 'SF_NOCACHE'
            static int __test_const_SF_NOCACHE_val = SF_NOCACHE;
```

This is because `SF_NOCACHE` was introduced very recently and hasn't made it to any releases yet. I added an exception to CI for it.

**Comment by alexcrichton** [comment]
Perhaps it could be elided for now? If it's so recent and not even released it may not be much use in practice anyway

**Comment by kamalmarhubi** [comment]
That makes sense, removed for now.

**Comment by kamalmarhubi** [comment]
For Apple I'm currently trying to get my olde laptoppe updated enough to be able to figure out what the generated code is saying.

**Comment by kamalmarhubi** [comment]
Ah, figured out the issue:

```diff
diff --git a/src/unix/bsd/apple/mod.rs b/src/unix/bsd/apple/mod.rs
index e59ae50..708b198 100644
--- a/src/unix/bsd/apple/mod.rs
+++ b/src/unix/bsd/apple/mod.rs
@@ -252,9 +252,9 @@ s! {
     }
 
     pub struct sf_hdtr {
-        pub headers: *const ::iovec,
+        pub headers: *mut ::iovec,
         pub hdr_cnt: ::c_int,
-        pub trailers: *const ::iovec,
+        pub trailers: *mut ::iovec,
         pub trl_cnt: ::c_int,
     }
 }
```

**Comment by kamalmarhubi** [comment]
@alexcrichton what's the release policy for libc? It'd be great to get a release soon with `sendfile(2)` and the `splice(2)` et al for Linux. :-)

**Comment by alexcrichton** [comment]
Looks like the signature may still be wrong for FreeBSD? And yeah I can push a new release soon as well, perhaps we can get the last few PRs right now merged first?

**Comment by kamalmarhubi** [comment]
Yeah one more attempt. I need to get that FreeBSD thing to run locally...

Re release, I mostly was wondering if it's days or weeks, so no worries about extra PRs.

**Comment by kamalmarhubi** [comment]
Ugh that one was missing full qualification of `c_int`.

**Pull request merged**